cpp/src/KIM_FilesystemPath.hpp

```diff
@@ -0,0 +1,203 @@
+  Path operator/(const Path & p) const;
+  friend bool operator<(const Path & lhs, const Path & rhs)
+  {
+    return lhs.path_ < rhs.path_;
+  }
+  friend bool operator==(const Path & lhs, const Path & rhs)
+  {
+    return lhs.path_ == rhs.path_;
+  }
+
+  // Concatenates the current path and the argument.
+  Path & concat(const std::string & p);
+
+  // Turns the path object into a conventional string, which can be passed to
+  // I/O functions.
+#ifdef KIM_API_USE_FILESYSTEM_LIBRARY
+  std::string string() const { return path_.string(); }
+#else
+  const std::string & string() const { return path_; }
+#endif
+
+  // Resets the path to an empty string.
+  void clear() { path_.clear(); }
+
+  // Returns whether this path is the empty string.
+  bool empty() const { return path_.empty(); }
+
+  // Returns the last component of the path.
+  Path filename() const;
+
+  // Removes a single filename component from the path.
+  Path & remove_filename();
+
+  // Converts all directory separators to the preferred directory separator of
+  // the current platform.
+  Path & make_preferred();
+
+  // Creates a new directory, including parent directories if necessary.
+  // It's not an error if the directory to be created already exists.
+  // Returns true on error.
+  bool make_directory() const;
+
+  // Deletes the contents of this path (if it is a directory) and the contents
+  // of all its subdirectories, recursively, then deletes the file path itself.
+  // Returns true on error.
+  bool remove_directory_recursive() const;
+
+  // Returns the list of subdirectories of this directory.
+  std::vector<Path> subdirectories() const;
+
+  // Checks whether the file or directory exists.
+  bool exists() const;
+
+  // Checks whether the path is relative.
+  bool is_relative() const;
+
+  // Returns the current working directory.
+  static Path current_path();
+
+  // Returns the user's home directory.
+  static Path home_path();
+
+  // Creates a new empty directory that can be used to write temporary files
+  // into. Returns an empty path on failure.
+  static Path create_temporary_directory(char const * const namePrefix);
+
+  // Performs stream output on the path (operator <<).
+  template<class CharT, class Traits>
+  friend std::basic_ostream<CharT, Traits> &
+  operator<<(std::basic_ostream<CharT, Traits> & os, const Path & p)
+  {
+    return os << p.path_;
+  }
+
+#ifndef KIM_API_USE_FILESYSTEM_LIBRARY
+  // Platform-dependent character for separating path entries.
+#ifndef _WIN32
+  static const std::string::value_type preferred_separator = '/';
+#else
+  static const std::string::value_type preferred_separator = '\\';
+#endif
+#endif
+
+ private:
+  // The internal path storage:
+#ifdef KIM_API_USE_FILESYSTEM_LIBRARY
+  std::filesystem::path path_;
+#else
+  std::string path_;
+#endif
+};
+
+class PathList : public std::vector<Path>
+{
+ public:
+  // Platform-dependent character for separating paths in the lists.
+#ifndef _WIN32
```

Comment:
This can be defined in the `.cpp` and the `#ifdef` avoided.

Reply:
Yes